### PR TITLE
Introduce module level parameters for zoracle

### LINKED ISSUE #490: Introduce module level parameters for zoracle
Follow Cosmos-SDK standard and introduce new parameters to `zoracle` module, which includes:
- `MaxDataSourceExecutableSize`: the maximum size of data source executable size, in bytes. Let's default this to 10 KB.
- `MaxOracleScriptCodeSize`: the maximum size of Owasm code, in bytes. Let's default this to 500 KB.
- `MaxCalldataSize`: the maximum size of calldata when invoking for oracle scripts or data sources.
- `MaxDataSourceCountPerRequest`: the maximum number of data sources a request can make. Let's default this to 16.
- `MaxRawDataReportSize`: the maximum size of raw data report per data source. Let's default this to 1 KB.

### PR DESCRIPTION
**Pull request by prin-r**
Fix #490

### CODE REVIEW

File: chain/x/zoracle/genesis.go [hunk 1]
```diff
@@ -6,12 +6,17 @@ import (
 	abci "github.com/tendermint/tendermint/abci/types"
 )
 
+// GenesisState is the bank state that must be provided at genesis.
```
Comment (sorawit) [inline chain/x/zoracle/genesis.go]: not bank

File: chain/x/zoracle/genesis.go [hunk 2]
```diff
@@ -29,8 +34,11 @@ func ValidateGenesis(data GenesisState) error {
 	return nil
 }
 
+// DefaultGenesisState returns a default genesis state
```
Comment (sorawit) [inline chain/x/zoracle/genesis.go]: fullstop

File: chain/x/zoracle/internal/types/params.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/x/params"
+)
+
+// Default parameter namespace
+const (
+	DefaultParamspace = ModuleName
+
+	// the maximum size of data source executable size, in bytes.
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go]: Capitalize the first character of sentences. Also fullstops.

File: chain/x/zoracle/internal/types/params.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		MaxDataSourceExecutableSize:  maxDataSourceExecutableSize,
+		MaxOracleScriptCodeSize:      maxOracleScriptCodeSize,
+		MaxCalldataSize:              maxCalldataSize,
+		MaxDataSourceCountPerRequest: maxDataSourceCountPerRequest,
+		MaxRawDataReportSize:         maxRawDataReportSize,
+	}
+}
+
+// String implements the stringer interface for Params
+func (p Params) String() string {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go:70]: What's the reason for all weird spaces before %d?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/genesis.go (hunk 1 above)
Comment (prin-r) [reply chain/x/zoracle/genesis.go]: done

File: chain/x/zoracle/genesis.go (hunk 2 above)
Comment (prin-r) [reply chain/x/zoracle/genesis.go]: done

File: chain/x/zoracle/internal/types/params.go (hunk 3 above)
Comment (prin-r) [reply chain/x/zoracle/internal/types/params.go]: done

File: chain/x/zoracle/internal/types/params.go (hunk 4 above)
Comment (prin-r) [reply chain/x/zoracle/internal/types/params.go:70]: done

File: chain/go.sum [hunk 5]
```diff
@@ -9,6 +9,7 @@ github.com/aead/siphash v1.0.1/go.mod h1:Nywa3cDsYNNK3gaciGTWPwHt0wlpNV15vwmswBA
 github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
 github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
 github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6/go.mod h1:grANhF5doyWs3UAsr3K4I6qtAmlQcZDesFNEHPZAzj8=
+github.com/bandprotocol/d3n v0.1.0 h1:GGjAfzvPGmdzedLFqKzp0EU+b5YVvshbzAern/XzZwo=
```
Comment (sorawit) [inline chain/go.sum]: Is this necessary?

File: chain/x/zoracle/genesis.go [hunk 6]
```diff
@@ -1,16 +1,22 @@
...
 import (
+	"github.com/bandprotocol/d3n/chain/x/zoracle/internal/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	abci "github.com/tendermint/tendermint/abci/types"
 )
 
+// GenesisState is the zoracle state that must be provided at genesis.
 type GenesisState struct {
 	// Scripts []types.StoredCode `json:"scripts"`
+	Params types.Params `json:"params" yaml:"params"` // inflation params
```
Comment (sorawit) [inline chain/x/zoracle/genesis.go]: inflation params???

File: chain/x/zoracle/genesis.go [hunk 7]
```diff
@@ -28,32 +34,25 @@ func ValidateGenesis(data GenesisState) error {
 	return nil
 }
 
+// DefaultGenesisState returns a default genesis state.
```
Comment (sorawit) [inline chain/x/zoracle/genesis.go]: the default

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 8]
```diff
@@ -13,18 +14,91 @@ type Keeper struct {
 	cdc           *codec.Codec
 	CoinKeeper    bank.Keeper
 	StakingKeeper staking.Keeper
+	Paramspace    params.Subspace
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: Let's call this `ParamSpace`. I think most of the sdk's packages use that name.

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 9]
```diff
@@ -13,18 +14,91 @@ type Keeper struct {
...
 }
 
 // NewKeeper creates a new zoracle Keeper instance.
-func NewKeeper(cdc *codec.Codec, key sdk.StoreKey, coinKeeper bank.Keeper, stakingKeeper staking.Keeper) Keeper {
+func NewKeeper(cdc *codec.Codec, key sdk.StoreKey, coinKeeper bank.Keeper, stakingKeeper staking.Keeper, paramspace params.Subspace) Keeper {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: use `paramSpace`

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 10]
```diff
@@ -13,18 +14,91 @@ type Keeper struct {
...
 	return Keeper{
 		storeKey:      key,
 		cdc:           cdc,
 		CoinKeeper:    coinKeeper,
 		StakingKeeper: stakingKeeper,
+		Paramspace:    paramspace.WithKeyTable(ParamKeyTable()),
 	}
 }
 
+// ParamKeyTable for zoracle module.
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: ParamKeyTable returns the parameter key table for zoracle module.

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 11]
```diff
@@ -13,18 +14,91 @@ type Keeper struct {
...
+func ParamKeyTable() params.KeyTable {
+	return params.NewKeyTable().RegisterParamSet(&types.Params{})
+}
+
+// Get MaxDataSourceExecutableSize.
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: Either remove all the comments or make them follow the godoc standard (function name + verb(s) + whole sentence).

// MaxDataSourceExecutableSize returns the maximum size of data source executable size in bytes.
// SetMaxDataSourceExecutableSize updates the MaxDataSourceExecutableSize parameter.

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 12]
```diff
@@ -13,18 +14,91 @@ type Keeper struct {
...
+	return
+}
+
+// Set SetMaxRawDataReportSize.
+func (keeper Keeper) SetMaxRawDataReportSize(ctx sdk.Context, value int64) {
+	keeper.Paramspace.Set(ctx, types.KeyMaxRawDataReportSize, value)
+}
+
+// Get all parameteras as types.Params.
+func (keeper Keeper) GetParams(ctx sdk.Context) types.Params {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go:82]: typo here. 
// GetParams returns all current parameters as a types.Params instance.

File: chain/x/zoracle/internal/types/params.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+	"github.com/cosmos/cosmos-sdk/x/params"
+)
+
+// Default parameter namespace.
+const (
+	DefaultParamspace = ModuleName
+
+	// The maximum size of data source executable size, in bytes.
+	// Default value is set to 10 kb.
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go]: use `kB` for Kilobyte. `kb` is KiloBit.

File: chain/x/zoracle/internal/types/params.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+		MaxDataSourceExecutableSize: %d
+		MaxOracleScriptCodeSize: %d
+		MaxCalldataSize: %d
+		MaxDataSourceCountPerRequest: %d
+		MaxRawDataReportSize: %d`, p.MaxDataSourceExecutableSize,
+		p.MaxOracleScriptCodeSize, p.MaxCalldataSize,
+		p.MaxDataSourceCountPerRequest, p.MaxRawDataReportSize)
+}
+
+// ParamSetPairs - Implements params.ParamSet.
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go]: // ParamSetPairs implements the params.ParamSet interface for Params.

File: chain/x/zoracle/internal/types/params.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+		{Key: KeyMaxDataSourceExecutableSize, Value: &p.MaxDataSourceExecutableSize},
+		{Key: KeyMaxOracleScriptCodeSize, Value: &p.MaxOracleScriptCodeSize},
+		{Key: KeyMaxCalldataSize, Value: &p.MaxCalldataSize},
+		{Key: KeyMaxDataSourceCountPerRequest, Value: &p.MaxDataSourceCountPerRequest},
+		{Key: KeyMaxRawDataReportSize, Value: &p.MaxRawDataReportSize},
+	}
+}
+
+// DefaultParams defines the parameters for this module.
+func DefaultParams() Params {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go:92]: defines the default parameters

**Review by sorawit — CHANGES_REQUESTED** [review]

**Comment by sorawit** [comment]
Logic looks good. Mostly comment on making the code looking more professional. @evilpeach @taobun please review and take note on how to retrieve parameters from keeper.

File: chain/x/zoracle/internal/keeper/keeper_test.go [hunk 16]
```diff
@@ -28,3 +29,61 @@ func TestGetNextRequestID(t *testing.T) {
 
 	require.Equal(t, uint64(4), keeper.GetRequestCount(ctx))
 }
+
+func TestGetSetMaxDataSourceExecutableSize(t *testing.T) {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper_test.go:33]: These tests are great!

File: chain/x/zoracle/internal/keeper/keeper.go (hunk 12 above)
Comment (prin-r) [reply chain/x/zoracle/internal/keeper/keeper.go:82]: done

File: chain/x/zoracle/internal/types/params.go (hunk 15 above)
Comment (prin-r) [reply chain/x/zoracle/internal/types/params.go:92]: done

File: chain/go.sum (hunk 5 above)
Comment (prin-r) [reply chain/go.sum]: done

File: chain/x/zoracle/genesis.go (hunk 7 above)
Comment (prin-r) [reply chain/x/zoracle/genesis.go]: done

File: chain/x/zoracle/internal/keeper/keeper.go (hunk 8 above)
Comment (prin-r) [reply chain/x/zoracle/internal/keeper/keeper.go]: done

File: chain/x/zoracle/internal/keeper/keeper.go (hunk 11 above)
Comment (prin-r) [reply chain/x/zoracle/internal/keeper/keeper.go]: done

File: chain/x/zoracle/internal/types/params.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+		MaxOracleScriptCodeSize:      maxOracleScriptCodeSize,
+		MaxCalldataSize:              maxCalldataSize,
+		MaxDataSourceCountPerRequest: maxDataSourceCountPerRequest,
+		MaxRawDataReportSize:         maxRawDataReportSize,
+	}
+}
+
+// String implements the stringer interface for Params.
+func (p Params) String() string {
+	return fmt.Sprintf(`Slashing Params:
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go]: Slashing?

File: chain/x/zoracle/internal/types/params.go [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+		MaxDataSourceExecutableSize: %d
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go]: I think the standard is two space.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/types/params.go [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+	return params.ParamSetPairs{
+		{Key: KeyMaxDataSourceExecutableSize, Value: &p.MaxDataSourceExecutableSize},
+		{Key: KeyMaxOracleScriptCodeSize, Value: &p.MaxOracleScriptCodeSize},
+		{Key: KeyMaxCalldataSize, Value: &p.MaxCalldataSize},
+		{Key: KeyMaxDataSourceCountPerRequest, Value: &p.MaxDataSourceCountPerRequest},
+		{Key: KeyMaxRawDataReportSize, Value: &p.MaxRawDataReportSize},
+	}
+}
+
+// Defines the default parameters.
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/params.go]: DefaultParams defines the default parameters.

Comment (sorawit) [reply chain/x/zoracle/internal/types/params.go]: https://blog.golang.org/godoc-documenting-go-code

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/alias.go [hunk 20]
```diff
@@ -6,9 +6,10 @@ import (
 )
 
 const (
-	ModuleName = types.ModuleName
-	RouterKey  = types.RouterKey
-	StoreKey   = types.StoreKey
+	ModuleName        = types.ModuleName
+	DefaultParamspace = types.DefaultParamspace
```
Comment (sorawit) [inline chain/x/zoracle/alias.go:10]: Why is this still using lowercase s? Let's do `DefaultParamSpace`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/alias.go (hunk 20 above)
Comment (sorawit) [reply chain/x/zoracle/alias.go:10]: ditto for all places. like 10 of them.

**Review by sorawit — APPROVED** [review]

File: chain/x/zoracle/alias.go (hunk 20 above)
Comment (sorawit) [reply chain/x/zoracle/alias.go:10]: Cosmos-SDK is very consistently inconsistent. They actually use `DefaultParamspace` in all places despite using `paramSpace` also in many places.

**Review by taobun — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**